File: 2021/day4/solution.py
from dataclasses import dataclass
from typing import List, Tuple

@dataclass
class Board:
    grid: List[List[str]]
    called: List[List[bool]]
    won: bool = False

    def mark(self, num: str) -> None:
        for r in range(len(self.grid)):
            for c in range(len(self.grid[0])):
                if self.grid[r][c] == num:
                    self.called[r][c] = True

    def has_bingo(self) -> bool:
        # check horizontal
        for r in range(len(self.grid)):
            all_true = True
            for c in range(len(self.grid[0])):
                if not self.called[r][c]:
                    all_true = False
                    break
            if all_true:
                self.won = True
                return True

        # check vertical
        for c in range(len(self.grid[0])):
            all_true = True
            for r in range(len(self.grid)):
                if not self.called[r][c]:
                    all_true = False
                    break
            if all_true:
                self.won = True
                return True

        return False

    def score(self, num: int) -> int:
        uncalled_sum = 0
        for r in range(len(self.grid)):
            for c in range(len(self.grid[0])):
                if not self.called[r][c]:
                    uncalled_sum += int(self.grid[r][c])
        
        return uncalled_sum * int(num)
    
@dataclass
class Callouts:
    nums: List[str]

def create_board(raw_board: List[List[str]]) -> Board:
    b = Board([], [])
    for row in raw_board:
        normalized_row = []
        curr = ""
        for c in row:
            if c.isnumeric():
                curr += c
            elif curr != "":
                normalized_row.append(curr)
                curr = ""
        normalized_row.append(curr)
        b.grid.append(normalized_row)
    
    for _ in range(len(b.grid)):
        row = []
        for _ in range(len(b.grid[0])):
            row.append(False)
        b.called.append(row)

    return b

def get_input() -> Tuple[Callouts, List[Board]]:
    groups = []
    local_group = []
    with open('input.txt', 'r') as file:
        for line in file:
            l = line.strip()
            if l == "":
                groups.append(local_group)
                local_group = []
            else:
                local_group.append(l)
        
        groups.append(local_group)

    callouts = Callouts(groups[0][0].split(','))
    boards = []
    for i in range(1, len(groups)):
        b = create_board(groups[i])
        boards.append(b)
    
    return (callouts, boards)

def get_first_bingo_score(callouts: Callouts, boards: List[Board]) -> int:
    for n in callouts.nums:
        for b in boards:
            b.mark(num=n)
            if b.has_bingo():
                return b.score(num=n)
    
    return -1

def get_last_winner(callouts: Callouts, boards: List[Board]) -> int:
    total_boards = len(boards)
    winners = 0

    for n in callouts.nums:
        for b in boards:
            if b.won:
                continue

            b.mark(num=n)
            if b.has_bingo():
                winners += 1
                if winners >= total_boards:
                    return b.score(num=n)
    
    return -1

callouts, boards = get_input()
p1 = get_first_bingo_score(callouts, boards)
print(p1)

callouts, boards = get_input()
p2 = get_last_winner(callouts, boards)
print(p2)

File: 2021/day4/solution.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

class Board {
private:
    vector<vector<bool>> called;
    vector<vector<int>> nums;
    bool won;
    
public:
    Board(const vector<vector<int>> nums) : nums(nums) {
        called = vector<vector<bool>>(5, vector<bool>(5, false));
    }

    void mark(const int number) {
        for (auto r = 0; r < nums.size(); r++) {
            for (auto c = 0; c < nums[r].size(); c++) {
                if (nums[r][c] == number) {
                    called[r][c] = true;
                    return;
                }
            }
        }
    }

    bool wins() {
        // check rows for match
        for (auto r = 0; r < nums.size(); r++) {
            bool match = true;
            for (auto c = 0; c < nums[r].size(); c++) {
                if (called[r][c] == false) {
                    match = false;
                    break;
                }
            }
            if (match) {
                won = true;
                return true;
            }
        }

        // check cols for match
        for (auto c = 0; c < nums[0].size(); c++) {
            bool match = true;
            for (auto r = 0; r < nums.size(); r++) {
                if (called[r][c] == false) {
                    match = false;
                    break;
                }
            }
            if (match) {
                won = true;
                return true;
            }
        }

        // check diagonals
        bool forward = called[0][0] && called[1][1] && called[2][2] && called[3][3] && called[4][4];
        bool backward = called[4][0] && called[3][1] && called[2][2] && called[1][3] && called[0][4];
        if (forward || backward) {
            won = true;
            return true;
        }
        return false;
    }

    int score(const int lastCalled) {
        int unmarkedSum = 0;
        for (auto r = 0; r < nums.size(); r++) {
            for (auto c = 0; c < nums[0].size(); c++) {
                if (!called[r][c]) {
                    unmarkedSum += nums[r][c];
                }
            }
        }
        return unmarkedSum * lastCalled;
    }

    bool has_won() {
        return won;
    }

    void print() {
        cout << "*****************" << endl;
        for (auto r = 0; r < nums.size(); r++) {
            string row = "";
            for (auto c = 0; c < nums[r].size(); c++) {
                if (c != 0) {
                    row.append(",");
                }
                row.append(to_string(nums[r][c]));
            }
            cout << row << endl;
        }
        cout << "*****************" << endl;
    }
};

vector<int> parseBingoNumbers(string numbers) {
    vector<int> bingoNumbers = {};
    stringstream ss(numbers);
    string item;
    while (getline(ss, item, ',')) {
        bingoNumbers.push_back(stoi(item));
    }
    return bingoNumbers;
}

vector<vector<int>> parseBoardNumbers(vector<string> rawBoard) {
    vector<vector<int>> result = {};
    for (auto i = 0; i < rawBoard.size(); i++) {
        vector<int> row = {};
        stringstream ss(rawBoard[i]);
        string item;
        while (ss >> item) {
            row.push_back(stoi(item));
        }
        result.push_back(row);
    }
    return result;
}

tuple<vector<int>, vector<Board>> parseInput(const string& filePath) {
    ifstream inputFile(filePath);
    vector<string> lines;
    if (!inputFile) {
        cerr << "Failed to open input file." << endl;
        return {};
    }
    string line;
    while (getline(inputFile, line)) {
        lines.push_back(line);
    }

    vector<int> bingoNumbers = parseBingoNumbers(lines[0]);
    
    vector<Board> boards = {};
    vector<string> local = {};
    for (auto i = 2; i < lines.size(); i++) {
        if (lines[i] == "") {
            boards.push_back(Board(parseBoardNumbers(local)));
            local = {};
        } else {
            local.push_back(lines[i]);
        }
    }

    return make_tuple(bingoNumbers, boards);
}

vector<int> scoreBoards(vector<int>& numbers, vector<Board> boards) {
    vector<int> scores = {};
    for (auto i = 0; i < numbers.size(); i++) {
        for (auto j = 0; j < boards.size(); j++) {
            if (boards[j].has_won() == false) {
                boards[j].mark(numbers[i]);
                if (boards[j].wins()) {
                    int localScore = boards[j].score(numbers[i]);
                    scores.push_back(localScore);
                }
            }
        }
    }
    return scores;
}



int main() {
    auto [numbers, boards] = parseInput("input.txt");
    vector<int> scores = scoreBoards(numbers, boards);

    cout << "Part 1: " + to_string(scores[0]) << endl;
    cout << "Part 2: " + to_string(scores[scores.size()-1]) << endl;
}